### PR TITLE
feat: Add node coloring and sizing to Dendrogram

### PR DESCRIPTION
**Pull request by jbrundage**
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
image

**Comment by jbrundage** [comment]
Now that I've pasted that screen shot I realize that I need to adjust the label positioning to match the radius of each node. I'll revisit this after addressing other issues/PRs.

File: packages/tree/src/Dendrogram.ts [hunk 1]
```diff
@@ -154,14 +155,34 @@ export class Dendrogram extends SVGZoomWidget {
         lines.exit().remove();
 
         //  Nodes  ---
-        const textOffsetX = this.circleRadius() + 2;
+        const textOffsetX = context.labelInNode() ? 0 : this.circleRadius() + 2;
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: Should be "this" not "context"

File: packages/tree/src/Dendrogram.ts [hunk 2]
```diff
@@ -154,14 +155,34 @@ export class Dendrogram extends SVGZoomWidget {
...
+                        let ret = 0;
+                        for (const child of node.children) {
+                            ret += _findSize(child);
+                        }
+                        return ret;
+                    }
+                    return 0;
+                }
+            })
+            .filter(n => n !== null)
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: When can n be null?

File: packages/tree/src/Dendrogram.ts [hunk 3]
```diff
@@ -193,9 +214,27 @@ export class Dendrogram extends SVGZoomWidget {
                 e.append("text");
             })
             ;
+        const min = Math.min(...sizes);
+        const max = Math.max(...sizes.slice(1));
+        const interpRadius = d3InterpolateNumber(this.minCircleRadius(), this.maxCircleRadius());
+        const interpColor = d3InterpolateRgb(this.lowerColor(), this.upperColor());
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: interpolateHsl gives a more natural interpolated color than interpolateRgb

File: packages/tree/src/Dendrogram.ts [hunk 4]
```diff
@@ -265,11 +307,36 @@ Dendrogram.prototype.implements(ITree.prototype);
...
 Dendrogram.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette", null, { tags: ["Intermediate", "Shared"] });
 Dendrogram.prototype.publish("mappings", [], "propertyArray", "Source Columns", null, { autoExpand: DendrogramColumn });
 
+Dendrogram.prototype.publish("labelInNode", false, "boolean", "If true, a node's label will be placed within its shape");
 Dendrogram.prototype.publish("circleRadius", 4.5, "number", "Text offset from circle");
+Dendrogram.prototype.publish("interpolateSizeToRadius", false, "boolean", "If true, minCircleRadius and maxCircleRadius reflect the size property of a data node");
+Dendrogram.prototype.publish("minCircleRadius", 4.5, "number", "Minimum radius of circle (based on size and only when interpolateSizeToRadius is true)", null, { disable: w => w.interpolateSizeToRadius() === false });
+Dendrogram.prototype.publish("maxCircleRadius", 14.5, "number", "Maximum radius of circle (based on size and only when interpolateSizeToRadius is true)", null, { disable: w => w.interpolateSizeToRadius() === false });
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: rename to "interpolateCircleRadiusMin" and "interpolateCircleRadiusMax"

File: packages/tree/src/Dendrogram.ts [hunk 5]
```diff
@@ -265,11 +307,36 @@ Dendrogram.prototype.implements(ITree.prototype);
...
+Dendrogram.prototype.publish("interpolateSizeToColor", false, "boolean", "If true, lowerColor and upperColor reflect the color of a data node");
+Dendrogram.prototype.publish("lowerColor", "#27ae60", "html-color", "Color used to represent the node with the lowest size property");
+Dendrogram.prototype.publish("upperColor", "#e74c3c", "html-color", "Color used to represent the node with the highest size property");
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: rename to "interpolateCircleColorLower" and "interpolateCircleColorUpper"

File: packages/tree/src/Dendrogram.ts [hunk 6]
```diff
@@ -265,11 +307,36 @@ Dendrogram.prototype.implements(ITree.prototype);
...
 Dendrogram.prototype.publish("mappings", [], "propertyArray", "Source Columns", null, { autoExpand: DendrogramColumn });
 
+Dendrogram.prototype.publish("labelInNode", false, "boolean", "If true, a node's label will be placed within its shape");
 Dendrogram.prototype.publish("circleRadius", 4.5, "number", "Text offset from circle");
+Dendrogram.prototype.publish("interpolateSizeToRadius", false, "boolean", "If true, minCircleRadius and maxCircleRadius reflect the size property of a data node");
+Dendrogram.prototype.publish("minCircleRadius", 4.5, "number", "Minimum radius of circle (based on size and only when interpolateSizeToRadius is true)", null, { disable: w => w.interpolateSizeToRadius() === false });
+Dendrogram.prototype.publish("maxCircleRadius", 14.5, "number", "Maximum radius of circle (based on size and only when interpolateSizeToRadius is true)", null, { disable: w => w.interpolateSizeToRadius() === false });
+Dendrogram.prototype.publish("interpolateSizeToColor", false, "boolean", "If true, lowerColor and upperColor reflect the color of a data node");
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: Rename to interpolateSizeToCircleColor

File: packages/tree/src/Dendrogram.ts (hunk 2 above)
Comment (jbrundage) [reply packages/tree/src/Dendrogram.ts]: Good catch, that was left over from an earlier version of that size mapping.

**Comment by GordonSmith** [comment]
@jbrundage bump

**Comment by GordonSmith** [comment]
@jbrundage bump II

**Comment by GordonSmith** [comment]
@jbrundage bump III

File: packages/tree/src/Dendrogram.ts [hunk 7]
```diff
@@ -154,14 +155,34 @@ export class Dendrogram extends SVGZoomWidget {
...
             if (context.radial()) {
                 return "rotate(" + (d.x - 90) + ")translate(" + d.y + ")";
             }
             return context.orientation() === "horizontal" ? "translate(" + d.y + "," + d.x + ")" : "translate(" + d.x + "," + d.y + ")";
         }
+        const sizes = dataNodes
+            .map(n => {
+                return _findSize(n);
+
+                function _findSize(node) {
```
Comment (GordonSmith) [inline packages/tree/src/Dendrogram.ts]: Remove inner function - its not adding any value that I can see?

Comment (jbrundage) [reply packages/tree/src/Dendrogram.ts]: It exists to allow recursion

Comment (GordonSmith) [reply packages/tree/src/Dendrogram.ts]: I missed that - apologies!

**Comment by GordonSmith** [comment]
@jbrundage bump IV

**Comment by jbrundage** [comment]
@GordonSmith this is revised with your suggested changes, but quite a bit of the code is of questionable quality. Mainly because certain features disable other features in unintuitive ways. 

For example: 
`labelInNode` places the nodes label in the center of the node's circle, but then the `interpolateCircleRadiusMin` can't be used as expected

**Pull request closed without merging**